File: src/gpu.ts
import { spawn } from "node:child_process";

export interface GpuSample {
  utilizationPercent: number;
  memoryUsedMib: number;
}

/**
 * Best-effort GPU utilization sampling via nvidia-smi. Never throws: a machine
 * with no NVIDIA GPU (most CI runners, most laptops) is the common case, not an
 * error case, so absence of nvidia-smi degrades to an empty sample list rather
 * than failing the whole `run` command.
 */
export async function sampleGpuOnce(): Promise<GpuSample[]> {
  return new Promise((resolve) => {
    const proc = spawn("nvidia-smi", [
      "--query-gpu=utilization.gpu,memory.used",
      "--format=csv,noheader,nounits"
    ]);
    let stdout = "";
    proc.stdout?.on("data", (chunk) => {
      stdout += chunk.toString();
    });
    proc.on("error", () => resolve([]));
    proc.on("close", (code) => {
      if (code !== 0 || !stdout.trim()) {
        resolve([]);
        return;
      }
      const samples = stdout
        .trim()
        .split("\n")
        .map((line) => {
          const [util, mem] = line.split(",").map((s) => Number.parseFloat(s.trim()));
          return { utilizationPercent: util, memoryUsedMib: mem };
        })
        .filter((s) => Number.isFinite(s.utilizationPercent) && Number.isFinite(s.memoryUsedMib));
      resolve(samples);
    });
  });
}

export function isGpuAvailable(): Promise<boolean> {
  return new Promise((resolve) => {
    const proc = spawn("nvidia-smi", ["-L"]);
    proc.on("error", () => resolve(false));
    proc.on("close", (code) => resolve(code === 0));
  });
}

File: python/src/computeledger/gpu.py
"""Best-effort GPU utilization sampling via ``nvidia-smi``, matching
``src/gpu.ts``. Never raises: a machine with no NVIDIA GPU (most CI
runners, most laptops) is the common case, not an error case, so absence of
``nvidia-smi`` degrades to an empty sample list rather than failing the
whole ``run`` command.
"""

from __future__ import annotations

import subprocess
from dataclasses import dataclass


@dataclass(frozen=True)
class GpuSample:
    utilization_percent: float
    memory_used_mib: float


def sample_gpu_once() -> list[GpuSample]:
    try:
        proc = subprocess.run(
            ["nvidia-smi", "--query-gpu=utilization.gpu,memory.used", "--format=csv,noheader,nounits"],
            capture_output=True,
            text=True,
            timeout=10,
        )
    except (FileNotFoundError, OSError, subprocess.TimeoutExpired):
        return []

    if proc.returncode != 0 or not proc.stdout.strip():
        return []

    samples: list[GpuSample] = []
    for line in proc.stdout.strip().split("\n"):
        parts = line.split(",")
        if len(parts) != 2:
            continue
        try:
            util = float(parts[0].strip())
            mem = float(parts[1].strip())
        except ValueError:
            continue
        samples.append(GpuSample(utilization_percent=util, memory_used_mib=mem))
    return samples


def is_gpu_available() -> bool:
    try:
        proc = subprocess.run(["nvidia-smi", "-L"], capture_output=True, timeout=10)
    except (FileNotFoundError, OSError, subprocess.TimeoutExpired):
        return False
    return proc.returncode == 0
